IdentityCore/src/cache/mac/MSIDMacTokenCache.m

```diff
@@ -22,39 +22,448 @@
+        return NO; \
+    } \
+}
+
+@interface MSIDMacTokenCache ()
+
+@property (nonatomic) NSMutableDictionary *cache;
+@property (nonatomic) dispatch_queue_t synchronizationQueue;
+
+@end
 
 @implementation MSIDMacTokenCache
 
-- (BOOL)setItemWithKey:(MSIDTokenCacheKey *)key
-                  item:(MSIDToken *)item
-            serializer:(id<MSIDTokenSerializer>)serializer
-               context:(id<MSIDRequestContext>)context
-                 error:(NSError **)error
+- (dispatch_queue_t)synchronizationQueue
 {
-    return NO;
+    if (!_synchronizationQueue) {
+        NSString *queueName = [NSString stringWithFormat:@"com.microsoft.msidmactokencache-%@", [NSUUID UUID].UUIDString];
+        _synchronizationQueue = dispatch_queue_create([queueName cStringUsingEncoding:NSASCIIStringEncoding], DISPATCH_QUEUE_CONCURRENT);
+    }
+    
+    return _synchronizationQueue;
+}
+
++ (MSIDMacTokenCache *)defaultCache
+{
+    static dispatch_once_t once;
+    static MSIDMacTokenCache *cache = nil;
+    
+    dispatch_once(&once, ^{
+        cache = [MSIDMacTokenCache new];
+    });
+    
+    return cache;
+}
+
+- (nullable NSData *)serialize
+{
+    if (!self.cache)
+    {
+        return nil;
+    }
+    
+    __block NSDictionary *cacheCopy = nil;
+    dispatch_sync(self.synchronizationQueue, ^{
+        cacheCopy = [self.cache mutableCopy];
+    });
+    
+    NSData *result = nil;
+    
+    // Using the dictionary @{ key : value } syntax here causes _cache to leak. Yay legacy runtime!
+    NSDictionary *wrapper = [NSDictionary dictionaryWithObjectsAndKeys:cacheCopy, @"tokenCache",@CURRENT_WRAPPER_CACHE_VERSION, @"version", nil];
+    
+    @try
+    {
+        NSMutableData *data = [NSMutableData data];
+        
+        NSKeyedArchiver *archiver = [[NSKeyedArchiver alloc] initForWritingWithMutableData:data];
+        // Maintain backward compatibility with ADAL.
+        [archiver setClassName:@"ADTokenCacheKey" forClass:MSIDTokenCacheKey.class];
+        [archiver setClassName:@"ADTokenCacheStoreItem" forClass:MSIDToken.class];
+        [archiver setClassName:@"ADUserInformation" forClass:MSIDUserInformation.class];
+        [archiver encodeObject:wrapper forKey:NSKeyedArchiveRootObjectKey];
+        [archiver finishEncoding];
+        
+        result = data;
+    }
+    @catch (id exception)
+    {
+        // This should be exceedingly rare as all of the objects in the cache we placed there.
+        MSID_LOG_ERROR(nil, @"Failed to serialize the cache!");
+    }
+    
+    return result;
+}
+
+- (BOOL)deserialize:(nullable NSData*)data
+              error:(NSError **)error
+{
+    NSDictionary *cache = nil;
+    __block BOOL result = NO;
+    
+    @try
+    {
+        NSKeyedUnarchiver *unarchiver = [[NSKeyedUnarchiver alloc] initForReadingWithData:data];
+        // Maintain backward compatibility with ADAL.
+        [unarchiver setClass:MSIDTokenCacheKey.class forClassName:@"ADTokenCacheKey"];
```

Comment:
nit: can this logic be part of JSON serializer?

Reply:
Let's add a Github issue to follow up in a separate PR.

Reply:
added #58